modules/uncertainty/R/ensemble.R

```diff
@@ -395,8 +395,8 @@ write.ensemble.configs <- function(defaults, ensemble.samples, settings, model,
                                             run.id = run.id
       )
       )
-      cat(run.id, file = file.path(settings$rundir, "runs.txt"), sep = "\n", append = TRUE)
-      
+      cat(format(run.id), file = file.path(settings$rundir, "runs.txt"), sep = "\n", append = TRUE)
```

Comment:
It's not clear to me what format is doing here -- is this to preserve the right representation of int64s? would explicit formatting arguments make that clearer?

Reply:
``` r
i <- bit64::as.integer64(9900000001)
cat(i)
#> 4.89125e-314
format(i)
#> [1] "9900000001"
```

My understanding is that there's a special `format.integer64` method that makes this printing work.

Reply:
I think the problem is that you cast it to a string with the format function:
```
> i <- 9900000001
> cat(i)
9.9e+09
> typeof(i)
[1] "double"
> typeof(bit64::as.integer64(i))
[1] "double"
> typeof(format(i))
[1] "character"
> options(scipen=12)
> cat(i)
9900000001
```
Would it make sense to set scipen to be 12 in PEcAn db package.

Reply:
@robkooper The problem is that `RPostgres` returns long integers as `bit64::integer64`, which is _not_ numeric. Here are some examples illustrating the issues with 64bit integers:

``` r
# Working with doubles is fine...
i <- 9900000001
class(i)
#> [1] "numeric"
typeof(i)
#> [1] "double"
cat(i, "\n")
#> 9.9e+09

# Working with integer64s is more complicated...
i64 <- bit64::as.integer64(i)
class(i64)
#> [1] "integer64"
typeof(i64)
#> [1] "double"

# Under the hood, this uses similar memory mapping to numerics, but
# with a very different meaning. As a result, when the class attribute
# is removed, integer64s are gibberish numerics... 
unclass(i64)
#> [1] 4.89125e-314

# ...which causes weird behavior when printing.
cat(i64, "\n")
#> 4.89125e-314

# This works as expected
cat(format(i64), "\n")
#> 9900000001

# Similarly, base R's `for` loop (and `lapply`) does not dispatch on
# `[[`, so it drops the class attribute, leading to problems
for (j in i64) print(j)
#> [1] 4.89125e-314
lapply(i64, function(x) x)
#> [[1]]
#> [1] 4.89125e-314

# Even `vapply` doesn't work.
vapply(i64, function(x) x, bit64::integer64(1))
#> [1] 4.89125e-314

# Have to either subset yourself...
for (jj in seq_along(i64)) print(i64[[jj]])
#> integer64
#> [1] 9900000001

# ...or use `purrr::map`
purrr::map(i64, ~.x)
#> [[1]]
#> integer64
#> [1] 9900000001

# ...but be careful with `unlist`, `purrr::simplify`, and friends,
# because they break it too!
l <- purrr::map(i64, ~.x)
l
#> [[1]]
#> integer64
#> [1] 9900000001
unlist(l)
#> [1] 4.89125e-314
purrr::simplify(l)
#> [1] 4.89125e-314
purrr::as_vector(l)
#> [1] 4.89125e-314
```

Reply:
Re: That last point -- the only reliable way I can see to `unlist` `integer64` values is via `do.call(c, i64)`.